From 7a75716c4c245ba0bcbe8161f75c2094976745dd Mon Sep 17 00:00:00 2001
From: Timon Sassor <timon.sassor@governikus.de>
Date: Wed, 18 Mar 2026 16:57:10 +0100
Subject: a11y Android: Listen to ACTION_SHOW_ON_SCREEN and pass it to
 QAccessible

Android sends ACTION_SHOW_ON_SCREEN if a "selection" or set focus of an
element outside the viewport is requested. This allows for positioning
of said element in the viewport. Since the action which is required to
position the element depends on the circumstances, the
AccessibleAttached showOnScreen signal was introduced to let the user
handle the positioning.

[ChangeLog][Accessibility][Android] React to the A11y
ACTION_SHOW_ON_SCREEN signal.

Change-Id: Ica99de5442af4d0a4fdd83bf1ad766c52cce03c4
Reviewed-by: Oliver Eftevaag <oliver.eftevaag@qt.io>
(cherry-picked from commit 2cf8cec9b79a12f1ab69b43d0e89b0dc35258499)
---
 .../qt/android/QtAccessibilityDelegate.java   |  8 +++++-
 .../qt/android/QtNativeAccessibility.java     |  1 +
 src/gui/accessible/qaccessible.cpp            | 14 ++++++++++
 src/gui/accessible/qaccessible.h              |  1 +
 .../android/androidjniaccessibility.cpp       | 27 +++++++++++++++++++
 5 files changed, 50 insertions(+), 1 deletion(-)

diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
index 06fb0d1ee4af716c44586328c3b55c3a4fe10f68..d12f84ecdd1e21acf9d2d1c12b0e2852352e991e 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtAccessibilityDelegate.java
@@ -492,6 +492,8 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
         screenRect.offset(-parentScreenRect.left, -parentScreenRect.top);
         setBoundsInParent(node, screenRect);
 
+        node.addAction(AccessibilityNodeInfo.AccessibilityAction.ACTION_SHOW_ON_SCREEN);
+
         // Manage internal accessibility focus state.
         if (m_focusedVirtualViewId == virtualViewId) {
             node.setAccessibilityFocused(true);
@@ -532,6 +534,10 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
 
             boolean handled = false;
             //Log.i(TAG, "PERFORM ACTION: " + action + " on " + virtualViewId);
+            if (action == AccessibilityNodeInfo.AccessibilityAction.ACTION_SHOW_ON_SCREEN.getId()) {
+                handled = QtNativeAccessibility.showOnScreen(virtualViewId);
+                notifyScrolledEvent(virtualViewId);
+            }
             switch (action) {
                 case AccessibilityNodeInfo.ACTION_CLEAR_ACCESSIBILITY_FOCUS:
                     if (m_focusedVirtualViewId == virtualViewId) {
@@ -544,7 +550,7 @@ class QtAccessibilityDelegate extends View.AccessibilityDelegate
                     m_view.invalidate();
                     sendEventForVirtualViewId(virtualViewId,
                             AccessibilityEvent.TYPE_VIEW_ACCESSIBILITY_FOCUS_CLEARED);
-                    handled = true;
+                    handled |= true;
                     break;
                 default:
                     // Let the node provider handle focus for the view node.
diff --git x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
index 97d135f217bbf325ab93a2046d61d53d4a43961c..dcd346d32aca970487fc719d37947f4589d4d2a6 100644
--- x/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
+++ y/qtbase/src/android/jar/src/org/qtproject/qt/android/QtNativeAccessibility.java
@@ -20,6 +20,7 @@ class QtNativeAccessibility
     static native boolean focusAction(int objectId);
     static native boolean scrollForward(int objectId);
     static native boolean scrollBackward(int objectId);
+    static native boolean showOnScreen(int objectId);
 
     static native boolean populateNode(int objectId, AccessibilityNodeInfo node);
 }
diff --git x/qtbase/src/gui/accessible/qaccessible.cpp y/qtbase/src/gui/accessible/qaccessible.cpp
index f735507651d5028de9d123fe7bd9d5aa2e6b6062..edaf81965b073de72ba43d98b77ec3bfadd11606 100644
--- x/qtbase/src/gui/accessible/qaccessible.cpp
+++ y/qtbase/src/gui/accessible/qaccessible.cpp
@@ -2859,6 +2859,7 @@ QAccessibleTableInterface::~QAccessibleTableInterface()
     \row    \li \l increaseAction() \li increase the value of the accessible (e.g. spinbox)
     \row    \li \l pressAction()    \li press or click or activate the accessible (should correspond to clicking the object with the mouse)
     \row    \li \l setFocusAction() \li set the focus to this accessible
+    \row    \li \l showOnScreenAction() \li move this accessible into the viewport
     \row    \li \l showMenuAction() \li show a context menu, corresponds to right-clicks
     \endtable
 
@@ -2950,6 +2951,7 @@ struct QAccessibleActionStrings
         decreaseAction(QStringLiteral(QT_TRANSLATE_NOOP("QAccessibleActionInterface", "Decrease"))),
         showMenuAction(QStringLiteral(QT_TRANSLATE_NOOP("QAccessibleActionInterface", "ShowMenu"))),
         setFocusAction(QStringLiteral(QT_TRANSLATE_NOOP("QAccessibleActionInterface", "SetFocus"))),
+        showOnScreenAction(QStringLiteral(QT_TRANSLATE_NOOP("QAccessibleActionInterface", "Show on Screen"))),
         toggleAction(QStringLiteral(QT_TRANSLATE_NOOP("QAccessibleActionInterface", "Toggle"))),
         scrollLeftAction(QStringLiteral(QT_TRANSLATE_NOOP("QAccessibleActionInterface", "Scroll Left"))),
         scrollRightAction(QStringLiteral(QT_TRANSLATE_NOOP("QAccessibleActionInterface", "Scroll Right"))),
@@ -2964,6 +2966,7 @@ struct QAccessibleActionStrings
     const QString decreaseAction;
     const QString showMenuAction;
     const QString setFocusAction;
+    const QString showOnScreenAction;
     const QString toggleAction;
     const QString scrollLeftAction;
     const QString scrollRightAction;
@@ -2984,6 +2987,8 @@ struct QAccessibleActionStrings
             return QAccessibleActionInterface::tr("Shows the menu");
         else if (actionName == setFocusAction)
             return QAccessibleActionInterface::tr("Sets the focus");
+        else if (actionName == showOnScreenAction)
+            return QAccessibleActionInterface::tr("Moves the accessible into viewport");
         else if (actionName == toggleAction)
             return QAccessibleActionInterface::tr("Toggles the state");
         else if (actionName == scrollLeftAction)
@@ -3061,6 +3066,15 @@ const QString &QAccessibleActionInterface::setFocusAction()
     return accessibleActionStrings()->setFocusAction;
 }
 
+/*!
+    Returns the name of the show on screen default action.
+    \sa actionNames(), localizedActionName()
+  */
+const QString &QAccessibleActionInterface::showOnScreenAction()
+{
+    return accessibleActionStrings()->showOnScreenAction;
+}
+
 /*!
     Returns the name of the toggle default action.
     \sa actionNames(), localizedActionName()
diff --git x/qtbase/src/gui/accessible/qaccessible.h y/qtbase/src/gui/accessible/qaccessible.h
index cf294f69e406175febca51a94cbb2e615311adfd..9aa223ed8b3a969a7ceb00adbb3386022fb11897 100644
--- x/qtbase/src/gui/accessible/qaccessible.h
+++ y/qtbase/src/gui/accessible/qaccessible.h
@@ -242,6 +242,7 @@ public:
     static const QString &decreaseAction();
     static const QString &showMenuAction();
     static const QString &setFocusAction();
+    static const QString &showOnScreenAction();
     static const QString &toggleAction();
     static QString scrollLeftAction();
     static QString scrollRightAction();
diff --git x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
index ab6d17aaabdb0a250852132e0f9b1e34785ff2d2..c479662f386aab8e7ef80cc1244248676269085b 100644
--- x/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
+++ y/qtbase/src/plugins/platforms/android/androidjniaccessibility.cpp
@@ -413,6 +413,32 @@ namespace QtAndroidAccessibility
         return result && oldPosition != screenRect_helper(firstChildId, false);
     }
 
+    static bool showOnScreen_helper(int objectId)
+    {
+        QAccessibleInterface *iface = interfaceFromId(objectId);
+        if (!iface || !iface->isValid() || !iface->actionInterface())
+            return false;
+
+        const auto actionNames = iface->actionInterface()->actionNames();
+
+        if (actionNames.contains(QAccessibleActionInterface::showOnScreenAction())) {
+            invokeActionOnInterfaceInMainThread(iface->actionInterface(), QAccessibleActionInterface::showOnScreenAction());
+            return true;
+        }
+        return false;
+    }
+
+    static jboolean showOnScreen(JNIEnv */*env*/, jobject /*thiz*/, jint objectId)
+    {
+        bool result = false;
+        if (m_accessibilityContext) {
+            runInObjectContext(m_accessibilityContext, [objectId]() {
+                return showOnScreen_helper(objectId);
+            }, &result);
+        }
+        return result;
+    }
+
     static QString textFromValue(QAccessibleInterface *iface)
     {
         QString valueStr;
@@ -827,6 +853,7 @@ namespace QtAndroidAccessibility
         {"focusAction", "(I)Z", (void*)focusAction},
         {"scrollForward", "(I)Z", (void*)scrollForward},
         {"scrollBackward", "(I)Z", (void*)scrollBackward},
+        {"showOnScreen", "(I)Z", (void *)showOnScreen}
     };
 
 #define GET_AND_CHECK_STATIC_METHOD(VAR, CLASS, METHOD_NAME, METHOD_SIGNATURE) \
